### ISSUE TITLE
How does the CMEMS harvester determine if a dataset exists at the source?

### ISSUE THREAD

**Issue by kmbn** [NextGeoss/ckanext-nextgeossharvest#26]
The harvester follows the same basic workflow for all the sources it harvests from:
1. Create a URL based on the `start_date` in the config.
2. Make a request to that URL.
3. If the the request returns 200, create a dataset. If the request does not return 200, or if it takes too long and the harvester aborts the request, do not create a dataset.

The logic seems to be that if the product exists at the source, the request will return 200 and if the product does not exist at the source, the request will not return 200. But the workflow treats the result of an outage the same as the result of trying to access a product that doesn't exist.

Do the sources not return a specific response when trying to access a non-existent product (e.g., 404 if there's no product, but 500, etc. if the error has another cause)?

**Comment by joaandrade** [comment]
This CMEMS providers are basically services running daily since a specific date.

For example: http://cmems.isac.cnr.it/thredds/wms/METOFFICE-GLO-SST-L4-NRT-OBS-SST-V2?

It has data since 2007-01-01. The service has products for all days since that date so we do not know the answer when a product does not exists.

If we test a date before 2007-01-01 (or a future date), the answer is 400 - The value "2006-12-31T12:00:00.000Z" is not valid for the TIME dimension.

In conclusion, even for the service outages (I never experienced one) the products for the outage days will be added in the ftp.

**Comment by kmbn** [comment]
Right, requesting a layer for a date that's too early or too recent (or too far in the future) will result in a 400 "The value "{your_datetime_here}" is not valid for the TIME dimension' error. But we're only going to be requesting layers for valid dates (yesterday, if the harvester is running normally, or a date further in the past but still within the valid time range if the harvester is catching up on historical data). We should only encounter that response if something happens to the source and it stops updating.

If we're sure that products will always be added to the FTP, even if there's an outage, why do we need to request the thumbnail and check the status code at all? If the products are always added, we can just skip the test requests, since we're not getting any additional metadata from them, and create a new dataset each day by filling in the URL templates for the resources.

http://thredds.met.no/thredds/wms/sea_ice/SIW-OSISAF-GLO-SIT_SIE_SIC-OBS/ice_conc_south_aggregated?request=GetMap&layers=ice_conc&version=1.3.0&crs=CRS:84&bbox=-180,-90,180,0&WIDTH=800&HEIGHT=800&styles=boxfill/rainbow&format=image/png&time=2018-03-16T12:00:00.000Z

**Comment by joaandrade** [comment]
Well, in fact, we have products for all days since the past limit date. However, the service can fail in the future. It is also good to keep the thumbnails for the users that are checking the catalogue via the front-end so, the WMS request is a way to guarantee that the product is there. Until now we have products for all the days but we cannot guarantee that they were all added in the supposed day. The best approach would be:

1- Split the different sources creating 3 different harvesters and put the similar code in the base file (/lib)

2- The harvester will run only three times a day but starting in the date of the last product (similar to Sentinel)

3- The end date should always be "TODAY"

4- The harvester should be able to handle a time interval to catalogue data from outage days.
E.g. Today, we have a timeout query for a specific product and for the three daily runs. Tomorrow the harvester should be able to start considering the last date in the catalogue.

**Comment by kmbn** [comment]
Your four points are a good starting point for reworking the the harvesters. They definitely need to be split up. Let's keep the issues separate, though. There's going to be a lot to keep track of for this group of harvesters and for the others that will come.

As for this issue, I'm still not convinced that the thumbnail query does anything more than verify that the WMS is responsive at the moment it's queried.

We're treating a non-200 response as a gather error. If this were a typical harvester, that would mean that something happened that prevented us from harvesting a dataset. Something that prevented us from getting metadata that we need to create a dataset.

E.g., we query SciHub for yesterday's products and get a 500 error. We log the error and move on because we can't create any datasets for those products because we don't have any of their metadata.

In this case, though, we're not querying the WMS in order to get metadata. The WMS doesn't tell us anything new about the product. For example, if SciHub goes down, we can't create any datasets. If the WMS goes down, we can still create new datasets. It could go down for a month and we could still create new datasets. The thumbnail links might not work during that time, but the datasets and their metadata would still be valid. 

Some requests to http://thredds.met.no/thredds/ncss/sea_ice timed out earlier today. Later they succeeded. What if they timed out each time the harvester ran? We wouldn't know that there were no products. We'd just know that the server was running slow.

I think we're making the harvesting process more complicated than it needs to be.

**Comment by joaandrade** [comment]
Agreed. So we can just check if the product exists in the FTP. If yes, we can create the dataset and save the thumbnail link without testing it (since we know that all the products have a correspondent WMS thumbnail). I mean, in fact the failure or timeout of a WMS query does not mean that the netcdf fle does not exist. So we should verify if the netcdf exist and, if yes, create the dataset and include the thumbnail url (without the need of testing it).

**Comment by kmbn** [comment]
The beta has been released. When it's time to rework the data hub and the harvesters we can decide if this issue is worth revisiting.
